src/DeviceInput/deviceInputSystem.ts

```diff
@@ -0,0 +1,223 @@
+ * pointer device and one keyboard.
+ */
+export class DeviceInputSystem {
+    // Private Members
+    private _inputs: Map<string, Array<number>> = new Map();
+    private _onDeviceConnected: (deviceName: string) => void = () => { };
+    private _onDeviceDisconnected: (deviceName: string) => void = () => { };
+    private _keyboardActive: boolean = false;
+    private _mouseActive: boolean = false;
+    private _elementToAttachTo: Nullable<HTMLElement>;
+
+    /**
+     * Default Constructor
+     * @param elementToAttachTo - element to attach events to (usually canvas)
+     */
+    constructor(elementToAttachTo: Nullable<HTMLElement>) {
+        this._elementToAttachTo = elementToAttachTo;
+        this.handleKeyActions();
+        this.handleMouseActions();
+        this.handleGamepadActions();
+        this.updateDevices();
+    }
+
+    // Public functions
+    /**
+     * Checks for current device input value, given an id and input index
+     * @param deviceName Id of connected device
+     * @param inputIndex Index of device input
+     * @returns Current value of input
+     */
+    public pollInput(deviceName: string, inputIndex: number): number {
+        if (this._inputs.has(deviceName)) {
+            var device = this._inputs.get(deviceName);
+
+            if (device![inputIndex] != undefined) {
+                return device![inputIndex];
+            }
+            else {
+                throw `Unable to find input ${inputIndex} on device ${deviceName}`;
+            }
+        }
+        else {
+            throw `Unable to find device ${deviceName}`;
+        }
+    }
+
+    /**
+     * When a device is connected, perform user specified function
+     * @param callback Callback function to use when a device is connected
+     */
+    public onDeviceConnected(callback: (deviceName: string) => void): void {
+        this._onDeviceConnected = callback;
+    }
+
+    /**
+     * When a device is disconnected, perform user specified function
+     * @param callback Callback function to use when a device is disconnected
+     */
+    public onDeviceDisconnected(callback: (deviceName: string) => void): void {
+        this._onDeviceDisconnected = callback;
+    }
+
+    // Private functions
+    /**
+     * Add device and inputs to device map
+     * @param deviceName Assigned name of device (may be SN)
+     * @param numberOfInputs Number of input entries to create for given device
+     */
+    private registerDevice(deviceName: string, numberOfInputs: number) {
+        if (!this._inputs.has(deviceName)) {
+            var device: Array<number> = [];
+            for (var i = 0; i < numberOfInputs; i++) {
+                device.push(0);
+            }
+
+            this._inputs.set(deviceName, device);
+        }
+    }
+
+    /**
+     * Given a specific device name, remove that device from the device map
+     * @param deviceName Name of device to be removed
+     */
+    private deregisterDevice(deviceName: string) {

```

Comment:
deRegisterDevice ?

Reply:
It should just be one word for the sake of camelcasing though I can change it to _deRegisterDevice_ if it helps with readability.

Reply:
Lol you know english better than me ;D

Reply:
@PolygonalSun perhaps we want "unregisterDevice()"? I'm not sure if deregister is proper english... its a slight nit, but something to consider.

Reply:
Changed to unregister because I only saw the use of that word throughout the BJS codebase.